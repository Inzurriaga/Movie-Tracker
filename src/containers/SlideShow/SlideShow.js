import React, { Component } from "react"
import { ReactComponent as Arrow } from "../../images/arrow.svg"
import { Link } from "react-router-dom"
import PropTypes from "prop-types"
import { connect } from "react-redux"

export class SlideShow extends Component{
    constructor(){
        super();
        this.state = {
            slides: [],
            slidesPosition: 0,
            autoSlides: () => {}
        }
    }

    componentDidMount = () => {
        const slides = this.props.movies.filter((movie, index) => {
            return index < 4
        })
        const autoSlides = setInterval(this.changeSlidesPositionForward, 6000)
        this.setState({slides, autoSlides})
    }

    changeSlidesPositionForward = () => {
        const { slidesPosition, slides } = this.state
        const autoSlides = setInterval(this.changeSlidesPositionForward, 6000)
        let position = slidesPosition
        if(position < slides.length - 1) {
            position++
            clearInterval(this.state.autoSlides)
            this.setState({ slidesPosition: position, autoSlides })
        } else{
            clearInterval(this.state.autoSlides)
            this.setState({ slidesPosition: 0, autoSlides})
        }
    }

    changeSlidesPositionBack = () => {
        const { slidesPosition } = this.state
        const autoSlides = setInterval(this.changeSlidesPositionForward, 6000)
        let position = slidesPosition
        if( slidesPosition === 0 ){
            clearInterval(this.state.autoSlides)
            this.setState({ slidesPosition: 3, autoSlides})
        }else{
            position--
            clearInterval(this.state.autoSlides)
            this.setState({ slidesPosition: position, autoSlides })
        }
    }

    render = () => {
        const slideDisplay = this.state.slides.map((slide, index) => {
            let backgroundImage = {backgroundImage: `url( http://image.tmdb.org/t/p/original${slide.backdrop_path})`}
            if(index === this.state.slidesPosition) {
                return (
                    <Link className="slide" key={slide.id} to={`/movies/${slide.id}`}>
                        <div className="slide-image" style={backgroundImage} id={slide.id}>
                            <h3>{slide.title}</h3>
                        </div>
                    </Link>
                    )
            }else{
                return null;
            }
        })
        return(
            <div className="slideshow-section">
                <div className="slide-container">
                    {slideDisplay}
                    <Arrow className="arrow-one" onClick={this.changeSlidesPositionForward}/>
                    <Arrow className="arrow-two" onClick={this.changeSlidesPositionBack}/>
                </div>
            </div>
        )
    }

    componentWillUnmount(){
        clearInterval(this.state.autoSlides)
    }
}

SlideShow.propTypes = {
    movies: PropTypes.array.isRequired,
}

export const mapStateToProps = (state) => ({
    movies: state.movies
})

export default connect(mapStateToProps)(SlideShow)
