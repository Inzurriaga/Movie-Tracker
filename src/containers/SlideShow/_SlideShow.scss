.slide-image {
    background-size: contain;
    width: 500px;
    height: 250px;
}