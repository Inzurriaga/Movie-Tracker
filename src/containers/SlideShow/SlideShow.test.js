import React from "react";
import { shallow, mount } from "enzyme"
import { SlideShow, mapStateToProps } from "./SlideShow"

const mockMovies = [{namne:"hello"},{name: "what"},{name: "yes"}]

describe("SlideShow", () => {
    describe("SlideShow", () => {
        let wrapper;

        beforeEach(() => {
            wrapper = shallow(
                <SlideShow movies={mockMovies}/>
            )
        })

        it("should match snap shot", () => {
            expect(wrapper).toMatchSnapshot()
        })

        it("should have a default state", () => {
            expect(wrapper.state()).toEqual({
                slides: [
                    { "namne": "hello"}, {"name": "what"},{ "name": "yes"}],
                slidesPosition: 0,
                autoSlides: 6
            })
        })

        it("should increase slidesPosition when invoke", () => {
            expect(wrapper.state("slidesPosition")).toEqual(0)
            wrapper.instance().changeSlidesPositionForward()
            expect(wrapper.state("slidesPosition")).toEqual(1)
        })

        it("should decrease slidesPosition when invoke", () => {
            wrapper.instance().changeSlidesPositionForward()
            expect(wrapper.state("slidesPosition")).toEqual(1)
            wrapper.instance().changeSlidesPositionBack()
            expect(wrapper.state("slidesPosition")).toEqual(0)
        })

    })

    describe("mapStateToProps", () => {
      
        it("should return a object", () => {
            const mockData = {
                movies: [],
                genres: []
            }
            const expected = {
                movies: []
            }
            const mockProps = mapStateToProps(mockData)
            expect(mockProps).toEqual(expected)
        })
    })
})
