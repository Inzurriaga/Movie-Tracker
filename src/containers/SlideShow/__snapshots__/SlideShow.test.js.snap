// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlideShow SlideShow should match snap shot 1`] = `
<div
  className="slideshow-section"
>
  <div
    className="slide-container"
  >
    <Link
      className="slide"
      to="/movies/undefined"
    >
      <div
        className="slide-image"
        style={
          Object {
            "backgroundImage": "url( http://image.tmdb.org/t/p/originalundefined)",
          }
        }
      >
        <h3 />
      </div>
    </Link>
    <ForwardRef
      className="arrow-one"
      onClick={[Function]}
    />
    <ForwardRef
      className="arrow-two"
      onClick={[Function]}
    />
  </div>
</div>
`;
